[SYSTEMDS-73] Simple lineage deduplication of ForProgramBlock

bnyra opened the pull request:
Hey @mboehm7

I'm sorry for being late. Here is my PR regarding to SYSTEMDS-73 lineage deduplication.
The PR includes the following changes:
 - Dedup test cases
 - Computation of distinct control flows in `ForProgramBlock` (without nested loops, branches and function calls)
 - Runtime branch evaluation of `ForProgramBlock`
 - Tracing of chosen branches in `ForProgramBlock`
 - Decompress functionality of dedup `LineageItem`

bnyra committed:
[SYSTEMDS-73] Add dedup testcase

bnyra committed:
[SYSTEMDS-73] Add first implementation of dedup

bnyra committed:
[SYSTEMDS-73] Add computation of distinct paths for IfProgramBlocks

bnyra committed:
[SYSTEMDS-73] Add branch evaluation on runtime

bnyra committed:
[SYSTEMDS-73] Decompress logic for dedup lineage items

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/controlprogram/context/ExecutionContext.java:
```diff
@@ -69,7 +70,9 @@
-
+	protected BitSet _lastBranch = null;
+	protected int _branchCounter = 0;
+	
```
Please create a data structure (e.g., LineagePath) that groups these attributes and functionality. Then you can simply provide a getLineagePath and move this functionality into your package.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/controlprogram/context/ExecutionContext.java:
```diff
@@ -651,4 +654,37 @@ public void cleanupCacheableData(CacheableData<?> mo) {
+		_branchCounter = 0;
+	}
+	
+	public void setBranchPredicateValue(Boolean value){
```
Please avoid objects (i.e., Boolean vs boolean) in the APIs.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/controlprogram/context/ExecutionContext.java:
```diff
@@ -651,4 +654,37 @@ public void cleanupCacheableData(CacheableData<?> mo) {
+		if (_lastBranch != null){
+			if (value)
+				_lastBranch.set(_branchCounter);
+			_branchCounter++;
```
Maybe simplify to _lastBranch.set(_branchCounter++, value);

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/controlprogram/context/ExecutionContext.java:
```diff
@@ -651,4 +654,37 @@ public void cleanupCacheableData(CacheableData<?> mo) {
+	public Long getLastBranch(){
+		if (_lastBranch != null)
+			if (_lastBranch.toLongArray().length == 1)
+				return _lastBranch.toLongArray()[0];
```
Please raise an exception if the length exceeds 1

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/lineage/Lineage.java:
```diff
@@ -16,150 +16,65 @@
+		boolean ret = false;
+		if (_initDedupBlock != null)
+			ret = _initDedupBlock.getActiveMap().containsKey(variable.getName());
+		ret |= _globalLineages.containsKey(variable.getName());
```
maybe simplify to return _initDedupBlock != null ? line 53 : false || line 54.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/lineage/Lineage.java:
```diff
@@ -16,150 +16,65 @@
+		if (_initDedupBlock != null)
+			li = _initDedupBlock.getActiveMap().get(variable);
+		if (li == null)
+			li = _globalLineages.get(variable);
```
simplify to return _initDedupBlock != null ? line 61 : line 63.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/lineage/LineageDedupBlock.java:
```diff
@@ -0,0 +1,96 @@
+	private void traceIfBodyInstructions(IfProgramBlock ipb, ExecutionContext ec) {
+		// Add IfBody instructions to lower half of LineageMaps
+		if (ipb.getChildBlocksIfBody() != null && ipb.getChildBlocksIfBody().size() == 1) {
+			for (Map.Entry<Long, LineageMap> entry : _distinctPaths.entrySet()) {
```
Please allow for a sequence of blocks in the if body.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/lineage/LineageDedupBlock.java:
```diff
@@ -0,0 +1,96 @@
+	private void traceElseBodyInstructions(IfProgramBlock ipb, ExecutionContext ec) {
+		// Add ElseBody instructions to upper half of LineageMaps
+		if (ipb.getChildBlocksElseBody() != null && ipb.getChildBlocksElseBody().size() == 1) {
+			for (Map.Entry<Long, LineageMap> entry : _distinctPaths.entrySet()) {
```
Please allow for a sequence of blocks in the else body.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/lineage/LineageDedupUtils.java:
```diff
@@ -0,0 +1,27 @@
+		for (ProgramBlock pb : fpb.getChildBlocks()) {
+			//TODO: This kind of type checking is very bad!!!
+			if (pb instanceof WhileProgramBlock || pb instanceof FunctionProgramBlock || pb instanceof ForProgramBlock)
+				throw new DMLRuntimeException("Deduplication is not supported for nested while for or function calls!");
```
Yes, please simply add an abstract method isNested() to ProgramBlock which forces every derived class to implement it.

bnyra replied:
I did not get the idea behind the abstract method `isNested()`.
Right now, only `ProgramBlock` as concrete instance, and `IfProgramBlock` are allowed inside a `ForProgramBlock`.
I have to check against iff ProgramBlock or if IfProgramBlock. Is there a way to check a BaseClass if there is no further concrete sub class.

mboehm7 replied:
you can leave it as it is for now - I'll take care of the general refactoring that is overdue anyway.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/lineage/LineageItemUtils.java:
```diff
@@ -122,33 +128,33 @@ public static Data computeByLineage(LineageItem root) {
 		//NOTE: we generate instructions from hops (but without rewrites) to automatically
 		//handle execution types, rmvar instructions, and rewiring of inputs/outputs
-		switch( item.getType() ) {
+		switch (item.getType()) {
```
Please fix the warning on missing cases.

bnyra replied:
Which functionality should I implement here?
Is it sufficient to raise a RuntimeError if we try to deserialize a lineage trace including dedup items.
Otherwise I have to implement more logic for deserialize dedup items from different files and find the right one at runtime.

mboehm7 replied:
just add a default branch that throws a controlled exception indicating the non-implemented type.

mboehm7 commented on src/main/java/org/tugraz/sysds/utils/Explain.java:
```diff
@@ -353,6 +355,8 @@ public static String explainLineageItems( ArrayList<LineageItem> lis, int level
 
 	public static String explain( LineageItem li ) {
+		if (DMLScript.LINEAGE_DEDUP)
+			li = LineageItemUtils.rDecompress(li);
```
I would recommend to move this outside the explain, such that we can use the explain internally for both compressed and uncompressed lineage traces.

mboehm7 replied:
Also it would be great if the explain of deduplicated lineage traces would show the dictionary of dedup traces and the ids of these traces.

mboehm7 reviewed (commented):
Thanks @bnyra - that's a great start; I just have a couple of minor comments.

bnyra commented:
Hey @mboehm7 
thank you for your review so far.

I addressed:
 - src/main/java/org/tugraz/sysds/runtime/lineage/LineageItemUtils.java line: 152
 - src/main/java/org/tugraz/sysds/runtime/lineage/LineageDedupUtils.java line: 14

See above at specific comment.

Right now I'm working on adoption of `explain(LineageItem li)`

bnyra committed:
[SYSTEMDS-73] Minor simplification fixes

bnyra committed:
[SYSTEMDS-73] More advanced logic for dedup explain()

bnyra committed:
[SYSTEMDS-73] Fix import statement

bnyra commented:
Thank you for your input.

I addressed all your comments and adapt the logic for dedup explain()
 - Add id to dedup lineage trace
 - At write instruction write compressed and decompressed lineage traces to disk if exists.
 - Add lineage trace for every distinct path into the explain string

bnyra commented:
My latest commit still fails with error:
cannot find symbol
method getInstructions()
location: variable pb of type org.tugraz.sysds.runtime.controlprogram.ProgramBlock

I guess this is due to recent refactoring. 
However, I did not merged or rebased it in. 
@mboehm7 could you help me to solve this problem pls?

mboehm7 commented:
sure - I'll have a closer over this PR tomorrow and make the any remaining fixes during the merge. My refactoring was mostly syntactic, meaning the `getInstructions()` was now moved to `BasicProgramBlock ` in order to make clear that only last-level blocks can have instructions other than predicate instructions.

mboehm7 commented:
LGTM - thanks for the patch and sorry for the delay. During the merge I fixed the conflicts, some missing reset operations, and unnecessary imports.
